src/gpu/generic/sycl/rnn/rnn_utils.hpp

```diff
@@ -0,0 +1,420 @@
+            ws_bias_size;
+
+    dim_t ws_gates_offset;
+    dim_t ws_states_offset;
+    dim_t ws_bias_offset;
+
+    // Element size of each workspace part in bytes
+    dim_t ws_gates_elsz, ws_states_elsz, ws_bias_elsz;
+
+    dim_t n_iter_scratch_gates;
+    dim_t scratch_gates_size, scratch_gates_elsz, scratch_gates_ld;
+    dims_t local_ranges;
+
+    data_type_t acc_data_type;
+    data_type_t aux_data_type;
+    data_type_t input_data_type;
+    data_type_t output_data_type;
+    data_type_t src_data_type;
+    data_type_t dst_data_type;
+    data_type_t wei_layer_type;
+    data_type_t wei_iter_type;
+    data_type_t bias_data_type;
+};
+
+dim_t get_good_ld(
+        dim_t arch_ld, dim_t dim, dim_t sizeof_dt, bool ignore_assoc = false);
+void init_rnn_conf(
+        conf_t &rnn, const rnn_pd_t *rnn_pd, data_type_t acc_data_type);
+void set_rnn_conf(conf_t &rnn, const rnn_desc_t &rd);
+dim_t set_workspace_offsets(
+        const conf_t &rnn, dim_t &ws_gates_offset, dim_t &ws_h_state_offset);
+dim_t get_workspace_size(const conf_t &rnn);
+status_t set_weights_desc(memory_desc_t &weights_md, const conf_t &rnn);
+status_t set_good_strides(memory_desc_t &weights_md, format_tag_t tag);
+const memory_storage_t &get_storage(const memory_storage_t *storage);
+const memory_storage_t &get_storage(
+        const std::unique_ptr<memory_storage_t> &storage);
+
+struct data_helper_t {
+    static dim_t type_size(data_type_t d) {
+        return static_cast<dim_t>(types::data_type_size(d));
+    }
+};
+
+struct user_data_t : public data_helper_t {
+    using mst = memory_storage_t;
+    user_data_t(const mst &src_layer, const mst &wei_layer, const mst &wei_iter,
+            const mst &bias, const conf_t &conf, const rnn_offsets_t &offsets)
+        : src_layer_(src_layer)
+        , wei_layer_(wei_layer)
+        , wei_iter_(wei_iter)
+        , bias_(bias)
+        , conf_(conf)
+        , offsets_(offsets) {
+        // The packed restriction could be removed by using batched GEMM with
+        // appropriate strides.
+        assert(offsets_.src_layer[0] < INT_MAX
+                && offsets_.src_layer[1] < INT_MAX
+                && offsets_.src_layer[2] < INT_MAX);
+    }
+
+    dim_t normalized_iter(dim_t dir, dim_t iter_) const { return iter_; }
+
+    strides_t<3> src_layer_strides(dim_t dir) const {
+        auto ret = offsets_.src_layer;
+
+        // Use negative iterations stride for backwards iteration
+        ret[0] *= (0 == normalized_iter(dir, 0)) ? 1 : -1;
+        return ret;
+    }
+
+    const mst &wei_layer() const { return wei_layer_; }
+    std::unique_ptr<mst> wei_layer(dim_t lay, dim_t dir) const {
+
+        // wei_layer dimension order: layer, dir, src c, gate, dst c
+        dim_t t = type_size(conf_.wei_layer_type);
+        dim_t lay_stride = offsets_.weights_layer[0];
+        dim_t dir_stride = offsets_.weights_layer[1];
+        dim_t offset = (lay * lay_stride + dir * dir_stride) * t;
```

Comment:
FYI, if the `memory_descriptor_t` structures were stored in `user_data_t`,  rather than `memory_storage_t`, we could use [memory_desc_wrapper::off](https://github.com/oneapi-src/oneDNN/blob/988240bd02b373526a3dafdc9e64a0ef7b3755fc/src/common/memory_desc_wrapper.hpp#L533-L540) to perform offset calculations.

Reply:
Great suggestion, added the `memory_desc_wrapper` and removed offsets struct.

Reply:
Thanks! Looking at the change, you should be able to remove the `conf` structure as well, it seems like it is only used for accessing the data type, but the data type is contained in the `memory_desc_wrapper`. You can directly access the size via https://github.com/oneapi-src/oneDNN/blob/8ec32403ff86e06f6633c71d9369a89be542b1dd/src/common/memory_desc_wrapper.hpp#L138-L139. 

One minor note on naming style: using a `_mdw` suffix is the standard method to denote a `memory_desc_wrapper`, rather than the `_wrap` used here.

Reply:
Suffix changed, thanks for pointing out.

Reply:
Additionally, removed the conf from the user_data_t in latest push `c499185`